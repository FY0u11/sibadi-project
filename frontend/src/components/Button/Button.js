import React from 'react'

const Button = ({ doAction, isButtonDisabled, disabled, children }) => {
  return (
    <button
      className="btn btn-dark"
      style={{ marginRight: '20px' }}
      onClick={doAction}
      disabled={disabled || isButtonDisabled}
    >
      {children}
    </button>
  )
}

export default Button
