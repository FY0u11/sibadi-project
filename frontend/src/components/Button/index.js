import React, { useState } from 'react'

import Button from './Button'

const ButtonContainer = ({ action, ...rest }) => {
  const [isButtonDisabled, setIsButtonDisabled] = useState(false)
  const doAction = async () => {
    const timer = setTimeout(() => setIsButtonDisabled(true), 200)
    await action()
    clearTimeout(timer)
    setIsButtonDisabled(false)
  }

  return <Button doAction={action && doAction} isButtonDisabled={isButtonDisabled} {...rest} />
}

export default ButtonContainer
