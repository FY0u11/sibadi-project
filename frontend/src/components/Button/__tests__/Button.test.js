import React from 'react'
import { render, cleanup, fireEvent, act } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Button from '../Button'
import ButtonContainer from '../'

afterEach(cleanup)

describe('Testing Button component', () => {
  it('It should be properly rendered', async () => {
    const { getByText, asFragment } = render(<Button>Click</Button>)
    expect(asFragment()).toMatchSnapshot()
    expect(getByText('Click')).toBeInTheDocument()
    expect(getByText('Click')).toBeEnabled()
  })
  it('It should be initially disabled', async () => {
    const { getByText } = render(<Button disabled>Click</Button>)
    expect(getByText('Click')).toBeDisabled()
  })
  it('It should become disabled while doing async task and enabled when it is done', async () => {
    const action = () => new Promise(resolve => setTimeout(resolve, 300))
    const button = render(<ButtonContainer action={action}>Click</ButtonContainer>).getByText(
      'Click'
    )
    await act(async () => {
      expect(button).toBeEnabled()
      fireEvent.click(button)
      await new Promise(resolve => setTimeout(resolve, 200))
      expect(button).toBeDisabled()
      await new Promise(resolve => setTimeout(resolve, 100))
      expect(button).toBeEnabled()
    })
  })
})
