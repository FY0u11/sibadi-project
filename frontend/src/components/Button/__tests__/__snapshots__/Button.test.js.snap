// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Button component It should be properly rendered 1`] = `
<DocumentFragment>
  <button
    class="btn btn-dark"
    style="margin-right: 20px;"
  >
    Click
  </button>
</DocumentFragment>
`;
